File: internal/gitx/git.go
package gitx

import (
	"bytes"
	"errors"
	"fmt"
	"os"
	"os/exec"
	"strconv"
	"strings"
)

var (
	ErrNotARepo = errors.New("not a git repository")

	ErrGitMissing = errors.New("git executable not found on PATH")
)

type Stats struct {
	Files   int
	Added   int
	Removed int
}

func run(args ...string) (string, error) {
	cmd := exec.Command("git", args...)
	var out, errBuf bytes.Buffer
	cmd.Stdout = &out
	cmd.Stderr = &errBuf
	if err := cmd.Run(); err != nil {
		if errors.Is(err, exec.ErrNotFound) {
			return "", ErrGitMissing
		}
		detail := strings.TrimSpace(errBuf.String())
		if detail == "" {
			detail = err.Error()
		}
		return "", fmt.Errorf("git %s: %s", strings.Join(args, " "), detail)
	}
	return out.String(), nil
}

func RepoRoot() (string, error) {
	out, err := run("rev-parse", "--show-toplevel")
	if err != nil {
		if errors.Is(err, ErrGitMissing) {
			return "", err
		}
		return "", ErrNotARepo
	}
	return strings.TrimSpace(out), nil
}

func StagedStats() (Stats, error) {
	out, err := run("diff", "--cached", "--numstat")
	if err != nil {
		return Stats{}, err
	}
	var s Stats
	for _, line := range strings.Split(out, "\n") {
		line = strings.TrimSpace(line)
		if line == "" {
			continue
		}
		parts := strings.SplitN(line, "\t", 3)
		if len(parts) < 3 {
			continue
		}
		s.Files++
		if n, err := strconv.Atoi(parts[0]); err == nil {
			s.Added += n
		}
		if n, err := strconv.Atoi(parts[1]); err == nil {
			s.Removed += n
		}
	}
	return s, nil
}

func excludeArgs(patterns []string) []string {
	args := []string{"--", "."}
	for _, p := range patterns {
		if p = strings.TrimSpace(p); p != "" {
			args = append(args, ":(exclude)"+p)
		}
	}
	if len(args) == 2 {
		return nil
	}
	return args
}

func StagedStat() (string, error) {
	return run("diff", "--cached", "--stat", "--no-color")
}

func StagedDiff(exclude []string) (string, error) {
	args := append([]string{"diff", "--cached", "--no-color"}, excludeArgs(exclude)...)
	return run(args...)
}

func stagedNames(exclude []string) ([]string, error) {
	args := append([]string{"diff", "--cached", "--name-only"}, excludeArgs(exclude)...)
	out, err := run(args...)
	if err != nil {
		return nil, err
	}
	var names []string
	for _, line := range strings.Split(out, "\n") {
		if line = strings.TrimSpace(line); line != "" {
			names = append(names, line)
		}
	}
	return names, nil
}

func ExcludedFiles(exclude []string) ([]string, error) {
	if len(exclude) == 0 {
		return nil, nil
	}
	all, err := stagedNames(nil)
	if err != nil {
		return nil, err
	}
	kept, err := stagedNames(exclude)
	if err != nil {
		return nil, err
	}
	keptSet := make(map[string]struct{}, len(kept))
	for _, name := range kept {
		keptSet[name] = struct{}{}
	}
	var dropped []string
	for _, name := range all {
		if _, ok := keptSet[name]; !ok {
			dropped = append(dropped, name)
		}
	}
	return dropped, nil
}

type CommitOptions struct {
	Edit   bool
	Verify bool
}

func Commit(message string, opts CommitOptions) (string, error) {
	f, err := os.CreateTemp("", "commitron-*.txt")
	if err != nil {
		return "", err
	}
	path := f.Name()
	defer os.Remove(path)

	if _, err := f.WriteString(message + "\n"); err != nil {
		f.Close()
		return "", err
	}
	if err := f.Close(); err != nil {
		return "", err
	}

	args := []string{"commit", "-F", path}
	if opts.Edit {
		args = append(args, "-e")
	}
	if !opts.Verify {
		args = append(args, "--no-verify")
	}

	cmd := exec.Command("git", args...)
	cmd.Stdin = os.Stdin

	cmd.Stdout = os.Stderr
	cmd.Stderr = os.Stderr
	if err := cmd.Run(); err != nil {
		return "", fmt.Errorf("git commit: %w", err)
	}

	out, err := run("rev-parse", "--short", "HEAD")
	return strings.TrimSpace(out), err
}

File: src/git.ts
import { spawnSync } from "node:child_process";
import { mkdtempSync, rmSync, writeFileSync } from "node:fs";
import { tmpdir } from "node:os";
import path from "node:path";

export class GitMissingError extends Error {
  constructor() {
    super("git executable not found on PATH");
  }
}

export class NotARepoError extends Error {
  constructor() {
    super("not a git repository");
  }
}

export interface Stats {
  files: number;
  added: number;
  removed: number;
}

export interface CommitOptions {
  edit: boolean;
  verify: boolean;
}

const maxBuffer = 256 * 1024 * 1024;

function run(args: string[], cwd?: string): string {
  const result = spawnSync("git", args, {
    cwd,
    encoding: "utf8",
    maxBuffer,
    windowsHide: true,
    stdio: ["ignore", "pipe", "pipe"],
  });
  if (result.error) {
    if ((result.error as NodeJS.ErrnoException).code === "ENOENT") {
      throw new GitMissingError();
    }
    throw result.error;
  }
  if (result.status !== 0) {
    const detail = result.stderr.trim() || describeExit(result.status, result.signal);
    throw new Error(`git ${args.join(" ")}: ${detail}`);
  }
  return result.stdout;
}

function describeExit(status: number | null, signal: NodeJS.Signals | null): string {
  return signal ? `killed by ${signal}` : `exit status ${status}`;
}

export function repoRoot(cwd?: string): string {
  try {
    return run(["rev-parse", "--show-toplevel"], cwd).trim();
  } catch (err) {
    if (err instanceof GitMissingError) {
      throw err;
    }
    throw new NotARepoError();
  }
}

export function stagedStats(cwd?: string): Stats {
  const stats: Stats = { files: 0, added: 0, removed: 0 };
  for (const raw of run(["diff", "--cached", "--numstat"], cwd).split("\n")) {
    const line = raw.trim();
    if (line === "") {
      continue;
    }
    const parts = line.split("\t");
    if (parts.length < 3) {
      continue;
    }
    stats.files++;
    stats.added += count(parts[0]);
    stats.removed += count(parts[1]);
  }
  return stats;
}

function count(field: string | undefined): number {
  const n = Number.parseInt(field ?? "", 10);
  return Number.isNaN(n) ? 0 : n;
}

function excludeArgs(patterns: string[]): string[] {
  const args = ["--", "."];
  for (const pattern of patterns) {
    const trimmed = pattern.trim();
    if (trimmed !== "") {
      args.push(`:(exclude)${trimmed}`);
    }
  }
  return args.length === 2 ? [] : args;
}

export function stagedStat(cwd?: string): string {
  return run(["diff", "--cached", "--stat", "--no-color"], cwd);
}

export function stagedDiff(exclude: string[], cwd?: string): string {
  return run(["diff", "--cached", "--no-color", ...excludeArgs(exclude)], cwd);
}

function stagedNames(exclude: string[], cwd?: string): string[] {
  return run(["diff", "--cached", "--name-only", ...excludeArgs(exclude)], cwd)
    .split("\n")
    .map((line) => line.trim())
    .filter((line) => line !== "");
}

export function excludedFiles(exclude: string[], cwd?: string): string[] {
  if (exclude.length === 0) {
    return [];
  }
  const kept = new Set(stagedNames(exclude, cwd));
  return stagedNames([], cwd).filter((name) => !kept.has(name));
}

export function commit(message: string, opts: CommitOptions, cwd?: string): string {
  const dir = mkdtempSync(path.join(tmpdir(), "commitron-"));
  try {
    const file = path.join(dir, "message.txt");
    writeFileSync(file, message + "\n");

    const args = ["commit", "-F", file];
    if (opts.edit) {
      args.push("-e");
    }
    if (!opts.verify) {
      args.push("--no-verify");
    }

    const result = spawnSync("git", args, { cwd, stdio: [0, 2, 2], windowsHide: true });
    if (result.error) {
      throw new Error(`git commit: ${result.error.message}`);
    }
    if (result.status !== 0) {
      throw new Error(`git commit: ${describeExit(result.status, result.signal)}`);
    }
  } finally {
    rmSync(dir, { recursive: true, force: true });
  }
  return run(["rev-parse", "--short", "HEAD"], cwd).trim();
}
